server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java

```diff
@@ -0,0 +1,394 @@
+
+            @Override
+            public PluginResponse read(StreamInput in) throws IOException {
+                return new PluginResponse(in);
+            }
+
+            @Override
+            public void handleResponse(PluginResponse response) {
+                for (DiscoveryExtensionNode extension : uninitializedExtensions) {
+                    if (extension.getName().equals(response.getName())) {
+                        extensions.add(extension);
+                        break;
+                    }
+                }
+            }
+
+            @Override
+            public void handleException(TransportException exp) {
+                logger.debug(new ParameterizedMessage("Plugin request failed"), exp);
+            }
+
+            @Override
+            public String executor() {
+                return ThreadPool.Names.GENERIC;
+            }
+        };
+        try {
+            transportService.connectToNode(extensionNode, true);
+            transportService.sendRequest(
+                extensionNode,
+                REQUEST_EXTENSION_ACTION_NAME,
+                new PluginRequest(transportService.getLocalNode(), new ArrayList<DiscoveryExtensionNode>(uninitializedExtensions)),
+                pluginResponseHandler
+            );
+        } catch (Exception e) {
+            logger.error(e.toString());
+        }
+    }
+
+    TransportResponse handleExtensionRequest(ExtensionRequest extensionRequest) throws Exception {
+        // Read enum
+        if (extensionRequest.getRequestType() == RequestType.REQUEST_EXTENSION_CLUSTER_STATE) {
+            ClusterStateResponse clusterStateResponse = new ClusterStateResponse(
+                clusterService.getClusterName(),
+                clusterService.state(),
+                false
+            );
+            return clusterStateResponse;
+        } else if (extensionRequest.getRequestType() == RequestType.REQUEST_EXTENSION_LOCAL_NODE) {
+            LocalNodeResponse localNodeResponse = new LocalNodeResponse(clusterService);
+            return localNodeResponse;
+        } else if (extensionRequest.getRequestType() == RequestType.REQUEST_EXTENSION_CLUSTER_SETTINGS) {
+            ClusterSettingsResponse clusterSettingsResponse = new ClusterSettingsResponse(clusterService);
+            return clusterSettingsResponse;
+        }
+        throw new Exception("Handler not present for the provided request");
+    }
+
+    public void onIndexModule(IndexModule indexModule) throws UnknownHostException {
+        for (DiscoveryNode extensionNode : uninitializedExtensions) {
+            onIndexModule(indexModule, extensionNode);
+        }
+    }
+
+    private void onIndexModule(IndexModule indexModule, DiscoveryNode extensionNode) throws UnknownHostException {
+        logger.info("onIndexModule index:" + indexModule.getIndex());
+        final CountDownLatch inProgressLatch = new CountDownLatch(1);
```

Comment:
The usage of `CountDownLatch` seems to be not the best option, there are at least several ways to approach response handlers chaining,  for example using `CompletableFuture<>`s, please let me know if you need help.

Reply:
I was able to implement half of the functionality with `CompletableFuture<>`, but when I try to invoke the `wait` method to mirror the latch `await` statements that we have, the forbidden API test fails.  Do you know a way other than the `wait` method to set a timeout like the latch does?